keps/sig-testing/20190605-presubmitconfig-in-tested-repo.md

```diff
@@ -0,0 +1,259 @@
+   * [Someone could accidentally delete all presubmits in prow.yaml and have their PR instantly merged](#someone-could-accidentally-delete-all-presubmits-in-prowyaml-and-have-their-pr-instantly-merged)
+* [Implementation History](#implementation-history)
+
+## Release Signoff Checklist
+
+**ACTION REQUIRED:** In order to merge code into a release, there must be an issue in [kubernetes/enhancements] referencing this KEP and targeting a release milestone **before [Enhancement Freeze](https://github.com/kubernetes/sig-release/tree/master/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core Kubernetes i.e., [kubernetes/kubernetes], we require the following Release Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These checklist items _must_ be updated for the enhancement to be released.
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+This document proposes to change the Prow presubmit handling to optionally version some or
+all presubmits in the same repository that also contains the code that is being tested.
+
+## Motivation
+
+Currently, all jobs for Prow are configured in a central `infra-config` repository that is
+in most cases distinct from the repositories whose code is being tested in presubmits. This
+poses severall challenges:
+
+* It is not possible to see in a Pull Request that introduces a new presubmit if that presubmit will
```

Comment:
For OpenShift, we are currently creating "rehearsal" jobs using the changed version of the job in question, so there are other useful workarounds that report job status to the PR changing the job.

Reply:
Could you point to some code I can look at? Sounds like a useful workaround

Reply:
https://github.com/openshift/ci-operator-prowgen/blob/master/cmd/pj-rehearse/main.go

We've thought about how to upstream it but in our impl we mixed information about the ProwJob with our executor (ci-operator) so it would be messy, and the security questions still loomed.